grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -731,6 +731,18 @@ class ProductService {
             addTagsToProduct(product, tags)
             addTagsToProduct(product, productProperties.tags)
 
+            ProductType defaultProductType = ProductType.defaultProductType.list()?.first();
+            // Throw an error for product type with empty code and product identifier that is not a default product type
+            if (product.productType?.id != defaultProductType?.id && !product.productType?.code && !product.productType?.productIdentifierFormat) {
+                throw new ValidationException("Could not save product '" + product.name + "'", product.errors)
```

Comment:
Instead of throwing an exception here you should just add to the product's errors object. 
```
if (product.productType?.id != defaultProductType?.id && !product.productType?.code && !product.productType?.productIdentifierFormat) {
    product.errors.reject("...")
}
```

The save below should fail due to the error and throw the exception. If for some reason it doesn't you can call validate on your own.
```
if (!product.validate() || !product.save(flush:true)) {
```

Reply:
Unfortunately doing `product.errors.reject("...")` and after that calling  `product.validate()` clears out previously assigned errors. I wasn't able to find any useful information on this topic but I assume that by calling  `product.validate()` it just resets `product.errors` and validates the object based on specified constrains in the class.

So instead I moved my validation logic into `validateProducts(csv)` _(where it probably should have been in the first place)_ and everything seems to work well there.

Reply:
I still see the exception being thrown here. 

https://github.com/openboxes/openboxes/pull/3547/files#diff-c318547de40b750cb640136d100a6aee922f7aae15814cdc220741c3e33d4575R641

Reply:
Oh nevermind. I see what you did. But just for clarification ... validate should not clear product.errors so we should try to figure out how to do this properly at some point.